amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -110,10 +121,84 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
                 throw new IllegalArgumentException(
                         "Provided ProcessRecordsInput not created from the FanOutRecordsPublisher");
             }
-            currentSequenceNumber = ((FanoutRecordsRetrieved) recordsRetrieved).continuationSequenceNumber();
+            currentSequenceNumber = recordsRetrieved.batchSequenceNumber();
+        }
+    }
+
+    @Override
+    public void notify(RecordsRetrievedAck recordsRetrievedAck) {
+        synchronized (lockObject) {
+            boolean isNextEventScheduled = false;
+            CompletableFuture<RecordFlow> triggeringFlowFuture = new CompletableFuture<>();
+            try {
+                isNextEventScheduled = evictAckedEventAndScheduleNextEvent(recordsRetrievedAck, triggeringFlowFuture);
+            } catch (Throwable t) {
+                errorOccurred(triggeringFlowFuture.getNow(null), t);
+            }
+            // TODO : debug log on isNextEventScheduled
+            final RecordFlow triggeringFlow = triggeringFlowFuture.getNow(null);
+            if (triggeringFlow != null) {
+                updateAvailableQueueSpaceAndRequestUpstream(triggeringFlow);
+            }
         }
     }
 
+    private boolean evictAckedEventAndScheduleNextEvent(RecordsRetrievedAck recordsRetrievedAck,
+            CompletableFuture<RecordFlow> triggeringFlowFuture) {
+        boolean isNextEventScheduled = false;
+        // Remove the head of the queue on receiving the ack.
+        // Note : This does not block wait to retrieve an element.
+        RecordsRetrievedContext recordsRetrievedContext = recordsDeliveryQueue.poll();
+        // Check if the ack corresponds to the head of the delivery queue.
+        if (recordsRetrievedContext != null && recordsRetrievedContext.getRecordsRetrieved().batchUniqueIdentifier()
+                .equals(recordsRetrievedAck.batchUniqueIdentifier())) {
+            // Update current sequence number for the successfully delivered event.
+            currentSequenceNumber = recordsRetrievedAck.deliveredSequenceNumber();
+            // Update the triggering flow for post scheduling upstream request.
+            triggeringFlowFuture.complete(recordsRetrievedContext.getRecordFlow());
+            // Try scheduling the next event in the queue, if available.
+            if (recordsDeliveryQueue.peek() != null) {
+                subscriber.onNext(recordsDeliveryQueue.peek().getRecordsRetrieved());
```

Comment:
this seems to introduce a race at record delivery. This method can find find the event in the queue as the next event, and deliver to Subscriber. But the event that is found could be the only event (added right after removal of the one upon the notification handled here), and another onNext would already be scheduled by the  bufferCurrentEventAndScheduleIfRequired.

Reply:
Hmmm bufferCurrentEventAndScheduleIfRequired would block wait on the lockObject until this is completed. After scheduling delivery by evictAckedEventAndScheduleNextEvent, the bufferCurrentEventAndScheduleIfRequired would acquire the lock, enqueue the next event, will find there are 2 events and will leave it to evictAckedEventAndScheduleNextEvent schedule the next event upon ack for previous event.

Reply:
I think that's right, all existing paths seem to get a hold on the lock. However, I'm not sure if this diff is showing all the new code changes introduced in this PR. I left a separate comment on that. Let's sync up offline.